README.md
# FDM Connector

This OctoPrint plugin will make the initial connection to FDM Monster a breeze. In future it will help in setting up a tunnel connection and tracking filament usage.

Current feature(s):
- Auto-registration - send your OctoPrint connection parameters to FDM Monster safely, to make setting up printers a breeze.

Future features:
- Filament Pedometer - send filament usage data to FDM Monster, cooperating with the plugin SpoolManager without external database (Filament Manager is not our aim)
- Http Tunnel - connect to FDM Monster, to make connection to printers a breeze especially over docker, VPN, DMZ, VLAN, the cloud or other complex network setups.
- Single-sign-On - client-to-machine (C2M) and machine-to-machine (M2M) authentication removing the need for more than 1 set of credentials across the farm.

For more feature requests, bugs, or ideas please head over to https://github.com/fdm-monster/fdm-connector/discussions.

## Setup
**!! Work in progress !!**
This plugin is in alpha stage. It requires the plugin system on FDM Monster, which is not released yet.

Install via the bundled [Plugin Manager](https://docs.octoprint.org/en/master/bundledplugins/pluginmanager.html)
or manually using this URL:

    https://github.com/fdm-monster/fdm-connector/archive/main.zip

Please configure the plugin completely for one or more printers before checking FDM Monster.

## Configuration
**Warning - restoring from a OctoPrint backup can be cause of security weakness due to copied passwords. In the end we are not responsible for your choices, but we advise to change the `accessControl:salt` and to regenerate the main user account for each OctoPrint instance.**

### Configuration - auto-registration
Configuring the auto-registration properly can massively improve the steps you need to undertake to setup your farm.
- REQUIRED `fdm_host` the host to reach FDM Monster with (IP, localhost, domain name, etc)
- REQUIRED `fdm_port` the port to approach the FDM Monster server (number)
- REQUIRED `oidc_client_id` the client id to authenticate with the FDM Monster server using OpenID Connect (string)
- REQUIRED `oidc_client_secret` the client secret key to authenticate with the FDM Monster server using OpenID Connect (string)
- OPTIONAL `port_override` the announced port for how FDM Monster can reach OctoPrint later (default is taken from `server:port`, but this is not always correct in case of Docker or a proxy.)

We understand if you restore OctoPrint backups to install new OctoPrints. For that reason we've introduced two unique ID's (UUID).
- AUTOGENERATED `persistence_uuid` a unique identifier stored in the plugin folder in `device.json`, which is excluded from backups to prevent duplicate printers. Dont adjust this, if you dont understand it.
- AUTOGENERATED `device_uuid` a unique identifier stored in the `config.yaml` at startup. Dont adjust this, if you dont understand it.

Periodic updates
- OPTIONAL `ping` the time in seconds between each call to FDM Monster (default is 15 * 60, or 15 minutes)

The plugin will use `server:host` and `server:port` to give FDM Monster a handle to connect back to this OctoPrint. This is often incorrect, if your OctoPrint is behind a proxy, in a VM, UnRaid, a different device, DMZ, in a docker container or in a VPN.
At this moment this needs to be rectified in FDM Monster. Later we will allow more advanced ways to fix this, but for now we believe this plugin is going to make it much easier already.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.2.0]

### Added

### Changed
    - Renamed to FDM Connector
### Removed

### Fixed


## [0.1.0-rc1-build3]

### Added
    - Test configuration and announcing with coverage up to 92%.

### Changed

### Removed

### Fixed
    - Abstracted state variable and accessor keys


## [0.1.0-rc1-build2]

### Added

### Changed

### Removed

### Fixed
    - #10 README fixed about settings saved to `config.yaml`
    - #11 Plugin URL was incorrect
    - #12 Password field wasnt hidden


## [0.1.0-rc1]

### Added
    - Feature: add unique key `device_uuid` to `config.yaml` to recognize the device
    - Feature: store backup-excluded unique key `perrsistence_uuid` to data folder to see when backups are restored
    - Feature: announce the OctoPrint device automatically using OpenID Connect with `client_credentials` as auth flow.
    - Feature: add settings and navbar templates to OctoPrint


tests/utils.py
from random import choice
from string import ascii_uppercase

from fdm_connector import Config

mocked_host_intercepted = "https://fdm123asdasdasdasd.com"

def mock_settings_get(accessor):
    if accessor[0] == "fdm_host":
        return Config.default_fdm_host
    if accessor[0] == "fdm_port":
        return Config.default_fdm_port
    if accessor[0] == "ping":
        return Config.default_ping_secs
    return None


def mock_settings_custom(accessor):
    if accessor[0] == "fdm_host":
        return mocked_host_intercepted
    if accessor[0] == "fdm_port":
        return 443
    if accessor[0] == "ping":
        return 300
    if accessor[0] == "oidc_client_id":
        return "ValidAnnoyer123"
    if accessor[0] == "oidc_client_secret":
        return "ValidPawo321"
    return None


def mock_settings_global_get(accessor):
    if accessor[0] == "server" and accessor[1] == "host":
        return Config.default_octoprint_host
    return None


def mock_settings_get_int(accessor):
    if accessor[0] == "fdm_port":
        return Config.default_fdm_port
    if accessor[0] == "ping":
        return Config.default_ping_secs
    return None


def create_fake_at():
    return ''.join(choice(ascii_uppercase) for i in range(Config.access_token_length))


tests/test_plugin.py
import unittest

from fdm_connector import __plugin_load__


class TestPlugin(unittest.TestCase):
    def test_load(self):
        __plugin_load__()


tests/test_connection.py
import json
import unittest
import unittest.mock as mock

import pytest
from werkzeug.exceptions import BadRequest

from fdm_connector import FdmConnectorPlugin, State


class TestPluginConnection(unittest.TestCase):
    @classmethod
    def setUp(cls):
        cls.settings = mock.MagicMock()  # Replace or refine with set/get
        cls.logger = mock.MagicMock()

        cls.plugin = FdmConnectorPlugin()
        cls.plugin._settings = cls.settings
        cls.plugin._logger = cls.logger
        cls.plugin._logger.info = print
        cls.plugin._logger.error = print
        cls.plugin._ping_worker = dict()  # disable it
        cls.plugin._data_folder = "test_data/connection"
        cls.plugin._write_persisted_data = lambda *args: None

    def assert_state(self, state):
        assert self.plugin._state is state

    # This method will be used by the mock to replace requests.get
    def mocked_requests_get(*args, **kwargs):
        class MockResponse:
            def __init__(self, json_data, status_code, text):
                self.json_data = json_data
                self.status_code = status_code
                self.text = text

            def json(self):
                return self.json_data

        return MockResponse({"version": "test-version"}, 200, json.dumps({"version": "test-version"}))

    @mock.patch('requests.get', side_effect=mocked_requests_get)
    def test_fdm_connection_test(self, mocked_requests_get):
        """Call the FDM Monster connection test properly"""

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1"})

        with mock.patch("fdm_connector.request", m):
            # somefile.method_called_from_route()
            response = self.plugin.test_fdmmonster_connection()
            assert response["version"] == "test-version"

    def _assert_bad_request_parameter(self, exception_info, param):
        assert str(exception_info.value) == f"400 Bad Request: Expected '{param}' parameter"

    @mock.patch('requests.get', side_effect=mocked_requests_get)
    def test_fdm_connection_test_validation(self, mocked_requests_get):
        """Call the FDM Monster connection test with faulty input"""

        m = mock.MagicMock()
        m.data = json.dumps({})
        with mock.patch("fdm_connector.request", m):
            with pytest.raises(BadRequest) as e:
                self.plugin.test_fdmmonster_connection()
            self._assert_bad_request_parameter(e, "url")

    @mock.patch('requests.post', side_effect=mocked_requests_get)
    def test_fdm_openid_validation(self, mocked_requests_get):
        """Call the FDM Monster OpenID connection test with faulty input"""

        m = mock.MagicMock()
        m.data = json.dumps({})
        with mock.patch("fdm_connector.request", m):
            with pytest.raises(BadRequest) as e:
                self.plugin.test_fdmmonster_openid()
            self._assert_bad_request_parameter(e, "url")

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1"})
        with mock.patch("fdm_connector.request", m):
            with pytest.raises(BadRequest) as e:
                self.plugin.test_fdmmonster_openid()
            self._assert_bad_request_parameter(e, "client_id")

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1", "client_secret": "asd"})
        with mock.patch("fdm_connector.request", m):
            with pytest.raises(BadRequest) as e:
                self.plugin.test_fdmmonster_openid()
            self._assert_bad_request_parameter(e, "client_id")

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1", "client_id": "asd"})
        with mock.patch("fdm_connector.request", m):
            with pytest.raises(BadRequest) as e:
                self.plugin.test_fdmmonster_openid()
            self._assert_bad_request_parameter(e, "client_secret")

    # This method will be used by the mock to replace requests.get
    def mocked_openid_response_notfound(*args, **kwargs):
        class MockResponse:
            def __init__(self, status_code, text):
                self.status_code = status_code
                self.text = text

        return MockResponse({}, 404)

    @mock.patch('requests.post', side_effect=mocked_openid_response_notfound)
    def test_fdm_openid_bug_response(self, mocked_requests_get):
        """Call the FDM Monster OpenID connection test with a not found error"""

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1", "client_id": "asd", "client_secret": "ok"})
        with mock.patch("fdm_connector.request", m):
            self.assert_state(State.BOOT)
            self.plugin.test_fdmmonster_openid()
            self.assert_state(State.CRASHED)

    # This method will be used by the mock to replace requests.get or requests.post
    def mocked_openid_response_maximal(*args, **kwargs):
        class MockResponse:
            def __init__(self, status_code, text):
                self.status_code = status_code
                self.text = text

        return MockResponse(200, json.dumps({"access_token": "test-token", "expires_in": 600, "token_type": "Bearer",
                                             "scope": "openid profile email pincode bank_id"}))

    # This method will be used by the mock to replace requests.get or requests.post
    def mocked_openid_response_minimal(*args, **kwargs):
        class MockResponse:
            def __init__(self, status_code, text):
                self.status_code = status_code
                self.text = text

        return MockResponse(200,
                            json.dumps({"access_token": "test-token", "expires_in": 600}))

    @mock.patch('requests.post', side_effect=mocked_openid_response_maximal)
    def test_fdm_openid_success_maximal(self, mocked_requests_get):
        """Call the FDM Monster OpenID connection test properly with maximum property set"""

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1", "client_id": "asd", "client_secret": "ok"})
        with mock.patch("fdm_connector.request", m):
            self.assert_state(State.BOOT)
            self.plugin.test_fdmmonster_openid()
            self.assert_state(State.SUCCESS)

    @mock.patch('requests.post', side_effect=mocked_openid_response_minimal)
    def test_fdm_openid_success_minimal(self, mocked_requests_get):
        """Call the FDM Monster OpenID connection test properly with missing response properties 'scope' and
        'token_type' """

        m = mock.MagicMock()
        m.data = json.dumps({"url": "http://127.0.0.1", "client_id": "asd", "client_secret": "ok"})
        with mock.patch("fdm_connector.request", m):
            self.assert_state(State.BOOT)
            self.plugin.test_fdmmonster_openid()
            self.assert_state(State.SUCCESS)


fdm_connector/constants.py
class Errors:
    access_token_too_short = "The 'access_token' did not meet the expected length of 43 characters. Preventing "
    "announcement query to FDM Monster"
    access_token_not_saved = "Conditional error: 'access_token' was not saved properly. Please report a bug to the " \
                             "plugin developers. Aborting "
    base_url_not_provided = "The 'base_url' was not provided. Preventing announcement query to FDM Monster"
    openid_config_unset = "Error connecting to FDM Monster. 'oidc_client_id' or 'oidc_client_secret' not set"
    config_openid_missing = "Configuration error: 'oidc_client_id' or 'oidc_client_secret' not set"
    ping_setting_unset = "'ping' config value not set. Aborting"


class Keys:
    persistence_uuid_key = "persistence_uuid"
    device_uuid_key = "device_uuid"


class Config:
    access_token_length = 43
    uuid_length = 36
    persisted_data_file = "backup_excluded_data.json"
    default_octoprint_host = "http://127.0.0.1"
    default_fdm_host = "http://127.0.0.1"
    default_fdm_port = 4000
    default_ping_secs = 120


class State:
    BOOT = "boot"
    SUCCESS = "success"
    SLEEP = "sleep"
    CRASHED = "crashed"
    RETRY = "retry"


fdm_connector/static/js/hub_connector.js
/*
 * View model for FDM Connector
 *
 * Author: David Zwart
 * License: AGPLv3
 */
$(function () {
    function FdmConnectorViewModel(parameters) {
        const self = this;

        self.loginState = parameters[0];
        self.settings = parameters[1];

        // this will hold the Fdm URL currently saved
        // self.currentUrl = ko.observable();
        //
        //         // This will get called before the HelloWorldViewModel gets bound to the DOM, but after its
        // // dependencies have already been initialized. It is especially guaranteed that this method
        // // gets called _after_ the settings have been retrieved from the OctoPrint backend and thus
        // // the SettingsViewModel been properly populated.
        // self.onBeforeBinding = function() {
        //     self.currentUrl(self.settings.settings.plugins.fdm_connector.fdm_host());
        //     self.currentPort(self.settings.settings.plugins.fdm_connector.fdm_port());
        //
        //     console.log(self.currentUrl(), self.currentPort())
        // }

        function getCurrentProposedUrl() {
            const host = $("#settings-fdm_connector-fdm_host")[0];
            const port = $("#settings-fdm_connector-fdm_port")[0];

            const proposedUrl = new URL(host.value);
            proposedUrl.port = port.value;

            return proposedUrl;
        }

        function getClientIdAndSecret() {
            const client_id = $("#settings-oidc-client-id")[0].value;
            const client_secret = $("#settings-oidc-client-secret")[0].value;

            return {
                client_id,
                client_secret
            };
        }

        self.settings.openfdmmonster = function () {
            const currentUrl = getCurrentProposedUrl();

            window.open(currentUrl, '_blank');
        }

        self.settings.testUrlBackend = async function () {
            const loader = $("#connection-loader");
            const successBar = $("#connection-success");
            const failureBar = $("#connection-failed");
            const responseVersion = $("#response-version");

            const currentUrl = getCurrentProposedUrl();

            var fullUrl = PLUGIN_BASEURL + "fdm_connector/test_fdmmonster_connection";

            loader.show();
            successBar.hide();
            failureBar.hide();
            responseVersion[0].innerText = "unset";
            return await fetch(fullUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    url: currentUrl
                })
            })
                .then(async (response) => {
                    try {
                        if (response.status >= 300) {
                            successBar.hide();
                            failureBar.show();
                        } else {
                            successBar.show();
                            const body = await response.json();
                            responseVersion[0].innerText = body.version;
                        }
                    } catch (e) {
                        console.log("Error occurred while testing FDM Monster", e);
                    } finally {
                        loader.hide();
                    }
                });
        };

        self.settings.testCredentialsBackend = async function () {
            const loader = $("#credentials-loader");
            const successBar = $("#credentials-success");
            const failureBar = $("#credentials-failed");
            const bugBar = $("#credentials-bug");
            const responseState = $("#response-state");

            const currentUrl = getCurrentProposedUrl();

            var fullUrl = PLUGIN_BASEURL + "fdm_connector/test_fdmmonster_openid";

            loader.show();
            successBar.hide();
            failureBar.hide();
            responseState[0].innerText = "unset";
            return await fetch(fullUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    url: currentUrl,
                    ...getClientIdAndSecret()
                })
            })
                .then(async (response) => {
                    try {
                        if (response.status == 400) {
                            successBar.hide();
                            bugBar.show();
                            failureBar.hide();
                        } else if (response.status >= 300) {
                            successBar.hide();
                            bugBar.hide();
                            failureBar.show();
                        } else {
                            successBar.show();
                            const body = await response.json();
                            responseState[0].innerText = body.state;
                        }
                    } catch (e) {
                        console.log("Error occurred while testing FDM Monster", e);
                    } finally {
                        loader.hide();
                    }
                });
        };
    }

    /* view model class, parameters for constructor, container to bind to
     * Please see http://docs.octoprint.org/en/master/plugins/viewmodels.html#registering-custom-viewmodels for more details
     * and a full list of the available options.
     */
    OCTOPRINT_VIEWMODELS.push([
        FdmConnectorViewModel,
        // ViewModels your plugin depends on, e.g. loginStateViewModel, settingsViewModel, ...
        [
            "loginStateViewModel", "settingsViewModel"
        ],
        // Elements to bind to, e.g. #settings_plugin_fdm_connector, #tab_plugin_fdm_connector, ...
        [
            // "#settings_plugin_fdm_connector",
            // "#navbar_plugin_fdm_connector"
        ]
    ]);
});


fdm_connector/__init__.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals

import io
import json
import os
import uuid
from datetime import datetime
from urllib.parse import urljoin

import flask
import octoprint.plugin
import requests
from flask import request
from octoprint.util import RepeatedTimer

from fdm_connector.constants import Errors, State, Config, Keys


def is_docker():
    path = '/proc/self/cgroup'
    return (
            os.path.exists('/.dockerenv') or
            os.path.isfile(path) and any('docker' in line for line in open(path))
    )


fdm_announce_route = 'api/plugins/octoprint/announce'
fdm_access_token_route = 'api/plugins/oidc/token'
fdm_version_route = 'api/version'
requested_scopes = 'openid'


class FdmConnectorPlugin(
    octoprint.plugin.StartupPlugin,
    octoprint.plugin.TemplatePlugin,
    octoprint.plugin.ShutdownPlugin,
    octoprint.plugin.BlueprintPlugin,
    octoprint.plugin.SettingsPlugin,
    octoprint.plugin.AssetPlugin,
):
    def __init__(self):
        self._ping_worker = None
        # device UUID and OIDC opaque access_token + metadata
        self._persisted_data = dict()
        self._excluded_persistence_datapath = None
        self._state = State.BOOT

    def on_after_startup(self):
        if self._settings.get(["fdm_host"]) is None:
            self._settings.set(["fdm_host"], Config.default_fdm_host)
        if self._settings.get(["fdm_port"]) is None:
            self._settings.set(["fdm_port"], Config.default_fdm_port)
        self._get_device_uuid()
        self._start_periodic_check()

    def get_excluded_persistence_datapath(self):
        self._excluded_persistence_datapath = os.path.join(self.get_plugin_data_folder(),
                                                           Config.persisted_data_file)
        return self._excluded_persistence_datapath

    def get_template_vars(self):
        fdm_host = self._settings.get(["fdm_host"])
        fdm_port = self._settings.get(["fdm_port"])
        base_url = f"{fdm_host}:{fdm_port}"
        favicon = f"{base_url}/favicon.ico"
        return dict(url=base_url, of_favicon=favicon)

    def get_template_configs(self):
        return [
            dict(type="settings", custom_bindings=False),
            dict(type="navbar", custom_bindings=False)
        ]

    # TODO make http://https:// slash robust
    def get_settings_defaults(self):
        return {
            "fdm_host": None,  # Without adjustment this config value is OFTEN useless
            "fdm_port": None,  # Without adjustment this config value is OFTEN useless
            "port_override": None,  # Without adjustment this config value is SOMETIMES useless
            "device_uuid": None,  # Auto-generated and unique
            "oidc_client_id": None,  # Without adjustment this config value is ALWAYS useless
            "oidc_client_secret": None,  # Without adjustment this config value is ALWAYS useless
            "ping": Config.default_ping_secs
        }

    def get_settings_version(self):
        return 1

    def get_assets(self):
        # Define your plugin's asset files to automatically include in the
        # core UI here.
        return dict(
            js=["js/fdm_connector.js"],
            css=["css/fdm_connector.css"],
            less=["less/fdm_connector.less"]
        )

    def initialize(self):
        self._fetch_persisted_data()

    def _fetch_persisted_data(self):
        filepath = self.get_excluded_persistence_datapath()
        if os.path.exists(filepath):
            try:
                with io.open(filepath, "r", encoding="utf-8") as f:
                    persistence_file = f.read()
                    persistence_json = json.loads(persistence_file)
                    self._persisted_data = persistence_json
            except json.decoder.JSONDecodeError as e:
                self._logger.warning(
                    "FDM Connector persisted device Id file was of invalid format.")
                self._write_new_device_uuid(filepath)
        else:
            self._write_new_device_uuid(filepath)

    def _write_new_access_token(self, filepath, at_data):
        self._persisted_data["access_token"] = at_data["access_token"]
        self._persisted_data["expires_in"] = at_data["expires_in"]
        self._persisted_data["requested_at"] = int(datetime.utcnow().timestamp())
        if "token_type" in at_data.keys():
            self._persisted_data["token_type"] = at_data["token_type"]
        if "scope" in at_data.keys():
            self._persisted_data["scope"] = at_data["scope"]
        self._write_persisted_data(filepath)
        self._logger.info("FDM Connector persisted data file was updated (access_token)")

    def _write_new_device_uuid(self, filepath):
        persistence_uuid = str(uuid.uuid4())
        self._persisted_data[Keys.persistence_uuid_key] = persistence_uuid
        self._write_persisted_data(filepath)
        self._logger.info("FDM Connector persisted data file was updated (device_uuid).")

    def _write_persisted_data(self, filepath):
        with io.open(filepath, "w", encoding="utf-8") as f:
            f.write(json.dumps(self._persisted_data))

    def _get_device_uuid(self):
        device_uuid = self._settings.get([Keys.device_uuid_key])
        if device_uuid is None:
            device_uuid = str(uuid.uuid4())
            self._settings.set([Keys.device_uuid_key], device_uuid)
            self._settings.save()
        return device_uuid

    def get_update_information(self):
        # Define the configuration for your plugin to use with the Software Update
        # Plugin here. See https://docs.octoprint.org/en/master/bundledplugins/softwareupdate.html
        # for details.
        return dict(
            fdm_connector=dict(
                displayName="FDM Connector Plugin",
                displayVersion=self._plugin_version,

                # version check: github repository
                type="github_release",
                user="fdm-monster",
                repo="fdm-connector",
                current=self._plugin_version,

                # update method: pip
                pip="https://github.com/fdm-monster/fdm-connector/archive/{target_version}.zip"
            )
        )

    def _start_periodic_check(self):
        if self._ping_worker is None:
            ping_interval = self._settings.get_int(["ping"])
            if ping_interval:
                self._ping_worker = RepeatedTimer(
                    ping_interval, self._check_fdmmonster, run_first=True
                )
                self._ping_worker.start()
            else:
                return self._logger.error(Errors.ping_setting_unset)

    def _check_fdmmonster(self):
        fdm_host = self._settings.get(["fdm_host"])
        fdm_port = self._settings.get(["fdm_port"])

        if fdm_host is not None and fdm_port is not None:
            base_url = f"{fdm_host}:{fdm_port}"

            # OIDC client_credentials flow result
            access_token = self._persisted_data.get("access_token", None)
            requested_at = self._persisted_data.get("requested_at", None)
            expires = self._persisted_data.get("expires", None)

            # Token expiry check - prone to time desync
            is_expired = None
            if requested_at is not None and expires is not None:
                current_time = datetime.utcnow().timestamp()
                is_expired = current_time > requested_at + expires

            token_invalid = not access_token or is_expired

            if token_invalid:
                oidc_client_id = self._settings.get(["oidc_client_id"])
                oidc_client_secret = self._settings.get(["oidc_client_secret"])
                self._logger.info("Refreshing access_token as it was expired")
                success = self._query_access_token(base_url, oidc_client_id, oidc_client_secret)
                if not success:
                    self._state = State.CRASHED
                    return False
            else:
                # We skip querying the token
                self._state = State.SUCCESS

            if "access_token" not in self._persisted_data.keys():
                # Quite unlikely as we'd be crashed
                raise Exception(Errors.access_token_not_saved)

            at = self._persisted_data["access_token"]

            self._query_announcement(base_url, at)

        else:
            self._logger.error(Errors.openid_config_unset)
            self._state = State.CRASHED
            raise Exception(Errors.config_openid_missing)

    def _query_access_token(self, base_url, oidc_client_id, oidc_client_secret):
        if not oidc_client_id or not oidc_client_secret:
            self._logger.error("Configuration error: 'oidc_client_id' or 'oidc_client_secret' not set")
            self._state = State.CRASHED
            return False

        at_data = None
        try:
            data = {'grant_type': 'client_credentials', 'scope': requested_scopes}
            self._logger.info("Calling FDM Connector at URL: " + base_url)
            url = urljoin(base_url, fdm_access_token_route)
            response = requests.post(url, data=data,
                                     verify=False, allow_redirects=False, auth=(oidc_client_id, oidc_client_secret))
            self._logger.info(response.text)
            self._logger.info(response.status_code)
            at_data = json.loads(response.text)
        except requests.exceptions.ConnectionError:
            self._state = State.RETRY  # TODO apply this with a backoff scheme
            self._logger.error("ConnectionError: error sending access_token request to FDM Monster")
        except Exception as e:
            self._state = State.CRASHED
            self._logger.error(
                "Generic Exception: error requesting access_token request to FDM Monster. Exception: " + str(e))

        if at_data is not None:
            if "access_token" not in at_data.keys():
                raise Exception(
                    "Response error: 'access_token' not received. Check your FDM Monster server logs. Aborting")
            if "expires_in" not in at_data.keys() is None:
                raise Exception("Response error: 'expires_in' not received. Check your FDM Monster server logs. Aborting")

            # Saves to file and to this plugin instance self._persistence_data accordingly
            self._write_new_access_token(self.get_excluded_persistence_datapath(), at_data)
            self._state = State.SUCCESS
            return True
        else:
            self._state = State.CRASHED
            self._logger.error("Response error: access_token data response was empty. Aborting")

    def _query_announcement(self, base_url, access_token):
        if self._state != State.SUCCESS and self._state != State.SLEEP:
            self._logger.error("State error: tried to announce when state was not 'success'")

        if base_url is None:
            self._state = State.CRASHED
            raise Exception(Errors.base_url_not_provided)

        if len(access_token) < 43:
            self._state = State.CRASHED
            raise Exception(Errors.access_token_too_short)

        # Announced data
        octoprint_port = self._settings.get(["port_override"])
        octoprint_host = self._settings.global_get(["server", "host"])
        # TODO maybe let FDM Monster decide instead of swapping ourselves?
        if octoprint_port is None:
            # Risk of failure when behind proxy (docker, vm, vpn, rev-proxy)
            octoprint_port = self._settings.global_get(["server", "port"])

        try:
            # Data folder based
            self._fetch_persisted_data()
            # Config file based
            device_uuid = self._get_device_uuid()

            check_data = {
                "deviceUuid": device_uuid,
                "persistenceUuid": self._persisted_data["persistence_uuid"],
                "host": octoprint_host,
                "port": int(octoprint_port),
                "docker": bool(is_docker())
            }

            headers = {'Authorization': 'Bearer ' + access_token}
            url = urljoin(base_url, fdm_announce_route)
            response = requests.post(url, headers=headers, json=check_data)

            self._state = State.SLEEP
            self._logger.info(f"Done announcing to FDM Monster server ({response.status_code})")
            self._logger.info(response.text)
        except requests.exceptions.ConnectionError:
            self._state = State.CRASHED
            self._logger.error("ConnectionError: error sending announcement to FDM Monster")

    def _call_validator_abort(self, key):
        flask.abort(400, description=f"Expected '{key}' parameter")

    @staticmethod
    def additional_excludes_hook(excludes, *args, **kwargs):
        return [Config.persisted_data_file]

    @octoprint.plugin.BlueprintPlugin.route("/test_fdmmonster_connection", methods=["POST"])
    def test_fdmmonster_connection(self):
        input = json.loads(request.data)
        keys = ["url"]
        for key in keys:
            if key not in input:
                return self._call_validator_abort(key)

        proposed_url = input["url"]
        self._logger.info("Testing FDM Monster URL " + proposed_url)

        url = urljoin(proposed_url, fdm_version_route)
        response = requests.get(url)
        version_data = json.loads(response.text)

        self._logger.info("Version response from FDM Monster " + version_data["version"])

        return version_data

    @octoprint.plugin.BlueprintPlugin.route("/test_fdmmonster_openid", methods=["POST"])
    def test_fdmmonster_openid(self):
        input = json.loads(request.data)
        keys = ["url", "client_id", "client_secret"]
        for key in keys:
            if key not in input:
                return self._call_validator_abort(key)

        proposed_url = input["url"]
        oidc_client_id = input["client_id"]
        oidc_client_secret = input["client_secret"]
        self._query_access_token(proposed_url, oidc_client_id, oidc_client_secret)

        self._logger.info("Queried access_token from FDM Monster")

        return {
            "state": self._state,
        }


__plugin_name__ = "FDM Connector"
__plugin_version__ = "0.2.0"
__plugin_description__ = "The FDM Monster plugin for OctoPrint"
__plugin_pythoncompat__ = ">=3,<4"


def __plugin_load__():
    global __plugin_implementation__
    __plugin_implementation__ = FdmConnectorPlugin()

    global __plugin_hooks__
    __plugin_hooks__ = {
        "octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information,
        "octoprint.plugin.backup.additional_excludes": __plugin_implementation__.additional_excludes_hook
    }
